File: dotnet-desktop-guide/winforms/controls/snippets/how-to-make-thread-safe-calls/cs/FormThread.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace project
{
    public partial class FormThread : Form
    {
        public FormThread()
        {
            InitializeComponent();
        }

        //<Good>
        private void button1_Click(object sender, EventArgs e)
        {
            var threadParameters = new System.Threading.ThreadStart(delegate { WriteTextSafe("This text was set safely."); });
            var thread2 = new System.Threading.Thread(threadParameters);
            thread2.Start();
        }

        public void WriteTextSafe(string text)
        {
            if (textBox1.InvokeRequired)
            {
                // Call this same method but append THREAD2 to the text
                Action safeWrite = delegate { WriteTextSafe($"{text} (THREAD2)"); };
                textBox1.Invoke(safeWrite);
            }
            else
                textBox1.Text = text;
        }
        //</Good>

        private void textBox1_Enter(object sender, System.EventArgs e)
        {
            //if (!String.IsNullOrEmpty(textBox1.Text))
            //{
            //    textBox1.SelectionStart = 0;
            //    textBox1.SelectionLength = textBox1.Text.Length;
            //}
        }
    }
}

File: dotnet-desktop-guide/winforms/controls/snippets/how-to-make-thread-safe-calls/vb/FormThread.vb
Public Class FormThread
    '<Good>
    Private Sub Button1_Click(sender As Object, e As EventArgs) Handles Button1.Click

        Dim threadParameters As New System.Threading.ThreadStart(Sub()
                                                                     WriteTextSafe("This text was set safely.")
                                                                 End Sub)

        Dim thread2 As New System.Threading.Thread(threadParameters)
        thread2.Start()

    End Sub

    Private Sub WriteTextSafe(text As String)

        If (TextBox1.InvokeRequired) Then

            TextBox1.Invoke(Sub()
                                WriteTextSafe($"{text} (THREAD2)")
                            End Sub)

        Else
            TextBox1.Text = text
        End If

    End Sub
    '</Good>
End Class
